lib/Subscription/Manage.php

```diff
@@ -21,11 +22,42 @@ static function onSave() {
 
     if (!empty($subscriber_data['email']) && Subscriber::verifyToken($subscriber_data['email'], $token)) {
       if ($subscriber_data['email'] !== Pages::DEMO_EMAIL) {
-        $subscriber = Subscriber::createOrUpdate($subscriber_data);
-        $errors = $subscriber->getErrors();
+        $subscriber = Subscriber::createOrUpdate(static::filterOutEmptyMandatoryFields($subscriber_data));
+        $subscriber->getErrors();
       }
     }
 
     Url::redirectBack();
   }
+
+  private static function filterOutEmptyMandatoryFields(array $subscriber_data) {
+    $mandatory = self::getMandatory();
+    $filtered = [];
+    foreach ($subscriber_data as $name => $value) {
+      if (!isset($mandatory[$name]) || !empty(trim($value))) {
+        $filtered[$name] = $value;
+      }
+    }
+    return $filtered;
+  }
+
+  private static function getMandatory() {
+    $mandatory = [];
+    $required_custom_fields = CustomField::findMany();
+    foreach ($required_custom_fields as $custom_field) {
+      if (is_serialized($custom_field->params)) {
+        $params = unserialize($custom_field->params);
+      } else {
+        $params = $custom_field->params;
+      }
+      if (
+        is_array($params)
+        && isset($params['required'])
+        && $params['required']
+      ) {
+        $mandatory['cf_' . $custom_field->id] = 'cf_' . $custom_field->id;
```

Comment:
Why are you returning an associative array in which the key is the value?
Returning a list of values makes more sense.
if you don't want to use `in_array` above (because it's slow), you can do
```php
foreach ($mandatory as $name) {
  if (strlen(trim($subscriber_data[$name])) === 0) {
    unset($subscriber_data[$name]);
  }
}
```